src/Server/TCPHandler.cpp

```diff
@@ -901,7 +901,7 @@ namespace
 
 void TCPHandler::sendProfileEvents()
 {
-    if (client_tcp_protocol_version < DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS)
+    if (client_tcp_protocol_version < DBMS_MIN_PROTOCOL_VERSION_WITH_INCREMENTAL_PROFILE_EVENTS)
```

Comment:
This will break binary compatibility of the protocol.

If some client implements `DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS` (=54450) then it will assume that some profile events will be sent, however now it will be send only if the client's version is `DBMS_MIN_PROTOCOL_VERSION_WITH_INCREMENTAL_PROFILE_EVENTS`, which is greater then `DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS`.

Reply:
Did not dig into details yet, by... I'm ok if we will remove sending profile events from 21.11 and introduce it only in 21.12.

Reply:
The problem is that original pr (#28364 10487f7) already included into 21.11 stable release, so simply removing it is not backward compatible, and so as changing the format.

Consider the following example:

Right now, without this patch, the following is possible:
- client with `DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS` will interpret `ProfileEvents` packets
- server with `DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS` will send `ProfileEvents` packets

After this change:
- client with `DBMS_MIN_PROTOCOL_VERSION_WITH_INCREMENTAL_PROFILE_EVENTS` will not interpret `ProfileEvents` packets that are send by server with `DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS` and so will throw `UNKNOWN_PACKET_FROM_SERVER`
- client with `DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS` will interpret `ProfileEvents` packets from server with `DBMS_MIN_PROTOCOL_VERSION_WITH_INCREMENTAL_PROFILE_EVENTS` incorrectly

I think that server should send nothing for `DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS` (it is totally OK, but worth putting a good comment about this), and send correct increment profile event packet for `DBMS_MIN_PROTOCOL_VERSION_WITH_INCREMENTAL_PROFILE_EVENTS`, that way at least client/server will be compatible across versions.

How does this sounds?

Reply:
@azat This sounds reasonable, otherwise every interaction between server and client with two consequtive versions (21.11 and 21.12) will lead to an exception. But for me it is Ok. It will be a tradeoff between a complexity of code and it usefulness. If there were some problems with server-server interaction, than we would have do this, because otherwise rolling update would not be possible or would be so problematic.

Reply:
@azat This sounds reasonable, otherwise every interaction between server and client with two consequtive versions (21.11 and 21.12) will lead to an exception. But for me it is Ok. It will be a tradeoff between a complexity of code and it usefulness. If there were some problems with server-server interaction, than we would have do this, because otherwise rolling update would not be possible or would be so problematic.

Reply:
There are 2 cases:
1. Client with `DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS` and server with `DBMS_MIN_PROTOCOL_VERSION_WITH_INCREMENTAL_PROFILE_EVENTS`. In this case server won't send `ProfileEvents` packets. It doesn't break anything.
2. Client with `DBMS_MIN_PROTOCOL_VERSION_WITH_INCREMENTAL_PROFILE_EVENTS` and server with `DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS`. Server will send incremental data, and client will try to interpret it as an absolute value. Client should be able to handle this values, but it will show incorrect hardware utilization. It's not that bad because in 21.11 it's shown only under `--hardware-utilization` option.

In general, I'd prefer to check server version on the client side, but I don't know how to do it right now.

Reply:
> client with DBMS_MIN_PROTOCOL_VERSION_WITH_INCREMENTAL_PROFILE_EVENTS will not interpret ProfileEvents packets that are send by server with DBMS_MIN_PROTOCOL_VERSION_WITH_PROFILE_EVENTS and so will throw UNKNOWN_PACKET_FROM_SERVER

@novikd seems you are right, and everything should work in this case, since client (and internal server clients) already has code for handling `Protocol::Server::ProfileEvents` and it will not go away regardless of server revision.